HW2/P2/parallel_vector.pyx

```diff
@@ -18,98 +19,115 @@ cimport numpy as np
-        omp_init_lock(&(locks[idx]))
+	if not locks:
+		with gil:
+			raise MemoryError()
+	for idx in range(N):
+		omp_init_lock(&(locks[idx]))
+
+	return locks
 
-    return locks
 
 cdef void free_N_locks(int N, omp_lock_t *locks) nogil:
-    cdef int idx
+	cdef int idx
 
-    for idx in range(N):
-        omp_destroy_lock(&(locks[idx]))
+	for idx in range(N):
+		omp_destroy_lock(&(locks[idx]))
 
-    free(<void *> locks)
+	free(<void *> locks)
 
 
 ##################################################
 # Your code below
 ##################################################
-
+#np.ndarray[np.int32_t, ndim=1] 
 cpdef move_data_serial(np.int32_t[:] counts,
-                       np.int32_t[:] src,
-                       np.int32_t[:] dest,
-                       int repeat):
-   cdef:
-       int idx, r
+						np.int32_t[:] src,
+						np.int32_t[:] dest,
+						int repeat):
+	cdef:
+		int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+	assert src.size == dest.size, "Sizes of src and dest arrays must match"
+	with nogil:
+		for r in range(repeat):
+			for idx in range(src.shape[0]):
+				if counts[src[idx]] > 0:
+					counts[dest[idx]] += 1
+					counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
-                             np.int32_t[:] src,
-                             np.int32_t[:] dest,
-                             int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
-
+							 np.int32_t[:] src,
+							 np.int32_t[:] dest,
+							 int repeat):
+	cdef:
+		int idx, r
+		omp_lock_t *locks = get_N_locks(counts.shape[0])
+	
+	##########
+	# Your code here
+	# Use parallel.prange() and a lock for each element of counts to parallelize
+	# data movement.  Be sure to avoid deadlock, and double-locking.
+	##########
+	with nogil:
+		for r in range(repeat):
+			for idx in prange(src.shape[0],num_threads=4):
+				acquire(&locks[src[idx]])
+				if counts[src[idx]] > 0:	
+					counts[src[idx]] -= 1
+					release(&locks[src[idx]])
+					acquire(&locks[dest[idx]])
+					counts[dest[idx]] += 1
+					release(&locks[dest[idx]])
+				else:
+					#Release so there is no deadlock
+					release(&locks[src[idx]])
+		
+
+	free_N_locks(counts.shape[0], locks)
```

Comment:
Were you able to avoid deadlocks using this code? The only way I know of is to check if the source and destination locks have lower/higher indices respectively :(